### PR TITLE
Enhance get_signature

### PR DESCRIPTION
**Pull request by pipiche38**
Makes get_signature method returning more device informations such as 
Model Id, 
Manufacturer Name/Code, 
ProfileID  
DeviceId

This will ease getting essential information on the application side.

PS/ I'm thinking to return battery level when the device has a battery component.

### CODE REVIEW

**Comment by pipiche38** [comment]
Do not understand the last issue of travis !! Please let me know what is expected !

File: zigpy/device.py [hunk 1]
```diff
@@ -338,13 +338,26 @@ def __getitem__(self, key):
         return self.endpoints[key]
 
     def get_signature(self):
-        signature = {}
+        # return the device signature by providing essential device information
+        #    - Model Identifier ( Attribut 0x0005 of Basic Cluster 0x0000 )
+        #    - Manufacturer Name ( Attribut 0x0004 of Basic Cluster 0x0000 )
+        #    - Endpoint list
+        #        - Profile Id, Device Id, Cluster Out, Cluster In
+        signature = {"manufacturer_name": '', "manufacture_code": '', "model": ''}
```
Comment (Adminiuga) [inline zigpy/device.py:346]: Just keep it empty.

**Comment by Adminiuga** [comment]
It is expecting tests cases covering both scenarios when when manuf/model is present and when not.

File: zigpy/device.py [hunk 2]
```diff
@@ -338,13 +338,26 @@ def __getitem__(self, key):
...
+        if self._manufacturer is not None:
+            signature["manufacturer_name"] = self.manufacturer
+        if self.manufacturer_id is not None:
+            signature["manufacture_code"] = self.manufacturer_id
```
Comment (Adminiuga) [inline zigpy/device.py:350]: Instead of manufacturer_id is could be better to store the node descriptor, as it gives you more information about the device

**Review by Adminiuga — COMMENTED** [review]
Address the comments and it is good to go

**Comment by puddly** [comment]
Make sure to follow the contributing guide's development environment installation instructions closely: https://github.com/zigpy/zigpy/blob/dev/CONTRIBUTING.md#setup

Right now the CI is failing because `black` would reformat your edits a little bit, which would be done automatically with the `pre-commit` hook. You can run those same checks locally and receive the same output by running `tox`.

File: zigpy/device.py (hunk 2 above)
Comment (pipiche38) [reply zigpy/device.py:350]: Good point, I'll see how to put those information too

File: zigpy/device.py (hunk 1 above)
Comment (pipiche38) [reply zigpy/device.py:346]: what do you mean ? I should keep signature empty, and then initialize each element ?

Comment (Adminiuga) [reply zigpy/device.py:346]: yes. do `signature = {}` if model, manufacturer are `None` then they should be missing from the sig

File: zigpy/device.py [hunk 3]
```diff
@@ -338,13 +338,31 @@ def __getitem__(self, key):
         return self.endpoints[key]
 
     def get_signature(self):
-        signature = {}
+        # return the device signature by providing essential device information
+        #    - Model Identifier ( Attribut 0x0005 of Basic Cluster 0x0000 )
+        #    - Manufacturer Name ( Attribut 0x0004 of Basic Cluster 0x0000 )
+        #    - Endpoint list
+        #        - Profile Id, Device Id, Cluster Out, Cluster In
+        signature = { }
```
Comment (Adminiuga) [inline zigpy/device.py:346]:
```suggestion
        signature = {}
```

**Comment by pipiche38** [comment]
> Why not just `signature['node_descriptor'] = str(self._node_desc)` ? That would fully allow you to reconstruct NodeDescriptor to the last detail if you have to

Agree, but doesn't it make sense to have to reconstruct later when we can get directly the information!

**Comment by Adminiuga** [comment]
> > Why not just `signature['node_descriptor'] = str(self._node_desc)` ? That would fully allow you to reconstruct NodeDescriptor to the last detail if you have to
> 
> Agree, but doesn't it make sense to have to reconstruct later when we can get directly the information!

then you could do `signature['node_descriptor'] = self._node_desc.as_dict()`

**Comment by pipiche38** [comment]
> > This will ease getting essential information on the application side.
> 
> can you expand a little bit on this? What essential information are you looking for? and what's the "application side? -- I assume it's not zigpy application, is it?
> What's the use case?

By application, i mean the application using the zigpy library. In my case the plugin for Domoticz. In that context, I need to create widgets in order to have the User interacting with the device.
List of clusters are going to give me what are the capabilities like Switch, Level Control, Window Covering ....
The Profile and Device Id can give me a bit more information on the device type
Finally the Model Name, will help me to handle specific devices and their features. I could have a cluster 0x0006 but get multi events and there is no real way to automatic discover it.

Then in normal time, when having to update a widget based on a attribute update I need also to correlate to the device information.

So to summarize, if the application (the plugin for instance) get all information in a dictionary like get_signature does this is good and there is no need to manipulate other means.

**Comment by Adminiuga** [comment]
Hrm, Ok.
I mean you also have the list of clusters in `device.endpoints[endpoint_id].in_clusters`. Not much different then accessing the signature dict. 

It seems to me internally that signature is used only in quirks testing. Not sure if ZHA UI leverages the same method or not

File: zigpy/device.py [hunk 4]
```diff
@@ -338,13 +338,27 @@ def __getitem__(self, key):
...
+        #    - Manufacturer Name ( Attribut 0x0004 of Basic Cluster 0x0000 )
+        #    - Endpoint list
+        #        - Profile Id, Device Id, Cluster Out, Cluster In
         signature = {}
+        if self._manufacturer is not None:
+            signature["manufacturer_name"] = self.manufacturer
+        if self._model is not None:
+            signature["model"] = self._model
+        if self.node_desc.is_valid:
+            signature["node_descriptor"] = self._node_desc.as_dict()
```
Comment (Adminiuga) [inline zigpy/device.py:352]: My bad. it is supposed to be 
```suggestion
            signature["node_descriptor"] = self.node_desc.as_dict()
```

**Comment by pipiche38** [comment]
> Hrm, Ok.
> I mean you also have the list of clusters in `device.endpoints[endpoint_id].in_clusters`. Not much different then accessing the signature dict.
> 
> It seems to me internally that signature is used only in quirks testing. Not sure if ZHA UI leverages the same method or not

This is may be where I come from and how I structured the access to ZiGate in the ZiGate plugin but I really appreciate to have only one call to get all infos.

PS/ Unfortunately for now, it is hard to test, as the zigpy-zigate master/dev doesn't work with zigpy/dev

**Comment by Adminiuga** [comment]
> but I really appreciate to have only one call to get all infos.

Well, this is duplicating. To me, the `zigpy` is not really that high. and `get_signature()` feels more like a representation of the device, which is already there. if the actual app needs that data structured in one place, `zigpy.device.Device` instance was supposed to be that representation and  if the app needs another "presentation", then the app should do so upon  starting zigpy and iterating over all devices. 
And since we apparently already using it for ZHA, we'll keep it.

> fortunately for now, it is hard to test, as the zigpy-zigate master/dev doesn't work with zigpy/dev

What's broken?

**Comment by pipiche38** [comment]
Thanks for approving the PR
